linux/mcompile/keymap/mc_import_rules.cpp

```diff
@@ -54,54 +52,53 @@ void DeadKey::KMX_AddDeadKeyRow(KMX_WCHAR baseCharacter, KMX_WCHAR combinedChara
 
 bool DeadKey::KMX_ContainsBaseCharacter(KMX_WCHAR baseCharacter) {
   std::vector<KMX_WCHAR>::iterator it;
-  for(it=this->m_rgbasechar.begin(); it<m_rgbasechar.end(); it++) {
-    if(*it == baseCharacter) {
+  for (it = this->m_rgbasechar.begin(); it < m_rgbasechar.end(); it++) {
+    if (*it == baseCharacter) {
       return true;
     }
   }
   return false;
 }
 
-int KMX_ToUnicodeEx(guint keycode, PKMX_WCHAR pwszBuff, int shift_state_pos, int caps,GdkKeymap *keymap) {
-  GdkKeymapKey *maps;
-  guint *keyvals;
+int KMX_ToUnicodeEx(guint keycode, PKMX_WCHAR pwszBuff, int shift_state_pos, int caps, GdkKeymap* keymap) {
+  GdkKeymapKey* maps;
+  guint* keyvals;
   gint count;
 
   if (!gdk_keymap_get_entries_for_keycode(keymap, keycode, &maps, &keyvals, &count))
     return 0;
 
-  if (!(shift_state_pos <= count))
-   return 0;
-
-  if (!(keycode <= keycode_max))
+  if (!(ensureValidInputForKeyboardTranslation(shift_state_pos, count, keycode)))
     return 0;
```

Comment:
Ah, here's another leak:

```suggestion
  if (!ensureValidInputForKeyboardTranslation(shift_state_pos, count, keycode)) {
    g_free(keyvals);
    g_free(maps);  
    return 0;
  }
```

Reply:
OK-done